mountpoint-s3/src/prefetch/parquet_prefetch.rs

```diff
@@ -0,0 +1,181 @@
+            "Invalid Parquet file. Reported metadata length is shorter than expected".to_string(),
+        ));
+    }
+
+    let metadata = chunk_reader.get_bytes(file_size - footer_metadata_len as u64, metadata_len)?;
+    Ok(metadata)
+}
+
+struct ParquetS3ChunkReader<Client: ObjectClient> {
+    client: Arc<Client>,
+    bucket: String,
+    key: String,
+    etag: ETag,
+    size: u64,
+}
+
+impl<Client: ObjectClient + Send + Sync + 'static> ChunkReader for ParquetS3ChunkReader<Client> {
+    type T = S3ReadSeek<Client>;
+
+    fn get_read(&self, start: u64, length: usize) -> Result<Self::T, ParquetError> {
+        Ok(S3ReadSeek {
+            client: self.client.clone(),
+            bucket: self.bucket.clone(),
+            key: self.key.clone(),
+            etag: self.etag.clone(),
+            position: start,
+            size: self.size,
+            length,
+        })
+    }
+}
+
+struct S3ReadSeek<Client: ObjectClient> {
+    client: Arc<Client>,
+    bucket: String,
+    key: String,
+    etag: ETag,
+    position: u64,
+    size: u64,
+    length: usize,
+}
+
+impl<Client: ObjectClient + Send + Sync + 'static> Read for S3ReadSeek<Client> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        trace!("Entered here for reader s3seek??");
+        let range_start = self.position;
+        let range_end = (self.position + self.length.min(buf.len()) as u64).min(self.size);
+        let range: std::ops::Range<u64> = range_start..range_end;
+
+        let span = debug_span!("read", range = ?range);
+        let _enter = span.enter();
+
+        if range.is_empty() {
+            return Ok(0);
+        }
+        let expected_size = range.end - range.start;
+        let mut data = Vec::with_capacity(expected_size as usize);
+
+        let runtime = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        let read_len = runtime.block_on(async {
+            let get_object_result = self
+                .client
+                .get_object(&self.bucket, &self.key, Some(range), Some(self.etag.clone()))
+                .await
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+
+            pin_mut!(get_object_result);
+
+            while let Some(result) = get_object_result.next().await {
+                match result {
+                    Ok((_, body)) => {
+                        trace!(length = body.len(), "received part");
+                        metrics::counter!("s3.client.total_bytes", "type" => "read").increment(body.len() as u64);
+                        data.extend_from_slice(&body);
+                    }
+                    Err(e) => return Err(std::io::Error::new(std::io::ErrorKind::Other, e)),
+                }
+            }
+
+            trace!("data fetched");
+            Ok(data.len())
+        })?;
+
+        buf[..read_len].copy_from_slice(&data);
```

Comment:
usually it's better to write data right in to the provided buffer `buf`, avoiding redundantly copying bytes in memory and an extra allocation for intermediate buffer `data`

Reply:
Agreed, will do that instead